File: real_outdec.py

import pandas as pd
import numpy as np

def real_outdec(station, real_outlier_threshold):
    
    # Read the csv file
    data = pd.read_csv(f"data/labeled_{station}_pro.csv", sep=',', encoding='utf-8')

    # Convert the 'date' column to datetime type
    data['date'] = pd.to_datetime(data['date'])

    # Extract the date part from the datetime and create a new column 'day'
    data['day'] = data['date'].dt.date

    # Group the data by 'day' and calculate the average of the 'label' column within each group
    average_labels = data.groupby('day', sort=False)['label'].mean()

    # Apply thresholding operation
    average_labels = average_labels.apply(lambda x: 1 if x >= real_outlier_threshold else 0)
    
    outliers_dates = average_labels[average_labels == 1].index
    outliers_indexes = np.where(average_labels == 1)[0]

    # Return the resulting objects
    return outliers_indexes, outliers_dates

File: real-outdec.R
# This file extracts the real outliers

real_outdec <- function(station) {

    # Read the csv file
    data <- read.csv(paste("data/labeled_", station, "_pro_msa.csv", sep = ""), header = TRUE, sep = ",", stringsAsFactors = FALSE)

    # Convert the 'date' column to a proper date format
    data$date <- as.Date(data$date)

    # Group the data by 'date' and calculate the average of the 'label' column within each group
    average_label_per_day <- data %>%
        group_by(date) %>%
        summarize(average_label = mean(label)) %>%
        ungroup() %>%
        arrange(match(date, data$date))

    # Store the average values in a numeric vector
    average_labels <- average_label_per_day$average_label

    # Assign names to the numeric vector in the desired format
    names(average_labels) <- as.character(average_label_per_day$date)

    # Thresholding operation: values above 0.1 set to 1, rest set to 0
    average_labels <- ifelse(average_labels >= 0.1, 1, 0)

    # Get the index of the days labeled as 1
    outliers_indexes <- which(average_labels == 1)

    return(outliers_indexes)

}